### PR TITLE
Sending refactor

### PR DESCRIPTION
**Pull request by marcbradshaw**
Work in progress.

### CODE REVIEW

**Comment by marcbradshaw** [comment]
unexpected  changes in README

**Comment by msimerson** [comment]
> unexpected changes in README

dzil likes to do that. I may or may not have added README to .gitignore locally.

File: lib/Mail/DMARC/Report/Sender.pm
Comment (msimerson) [inline lib/Mail/DMARC/Report/Sender.pm]: Is this to coerce the returned type?

Comment (marcbradshaw) [reply lib/Mail/DMARC/Report/Sender.pm]: No, It was purely to aid debugging.
For info, I'm waiting for a change to Email::Sender::Transport::SMTP to be merged before merging this change.

**Comment by marcbradshaw** [comment]
rebased on top of the current master

**Comment by msimerson** [comment]
From the currently failing build test:

````
t/26.Report.Sender.t ......................... Can't locate Email/Sender/Transport/Test.pm in @INC (you may need to install the Email::Sender::Transport::Test module)
````

File: README.md
Comment (msimerson) [inline README.md:291]: Please revert all the changes in this file.

**Review by msimerson — CHANGES_REQUESTED** [review]
I submitted a PR against your branch which resolves all the issues except for the mail transport failure.

**Comment by msimerson** [comment]
````
t/26.Report.Sender.t ......................... 
1320# Subtest: method
1321using /tmp/b9DclANGof/auto/share/dist/Mail-DMARC/mail-dmarc.ini for mail-dmarc.ini
1322sleeping 5.....done.
1323    not ok 1 - 1 Email sent
1324    #   Failed test '1 Email sent'
1325    #   at t/26.Report.Sender.t line 113.
1326    #          got: '0'
1327    #     expected: '1'
1328    not ok 2 - Sent to correct address
1329    #   Failed test 'Sent to correct address'
1330    #   at t/26.Report.Sender.t line 114.
1331    #          got: undef
1332    #     expected: 'rua@fastmaildmarc.com'
1333    1..2
1334    # Looks like you failed 2 tests of 2.
1335not ok 1 - method
1336#   Failed test 'method'
1337#   at t/26.Report.Sender.t line 120.
1338Can't use an undefined value as a SCALAR reference at t/26.Report.Sender.t line 115.
1339# Tests were run but no plan was declared and done_testing() was not seen.
1340# Looks like your test exited with 255 just after 1.
1341Dubious, test returned 255 (wstat 65280, 0xff00)
1342Failed 1/1 subtests 
1343
1344Test Summary Report
1345-------------------
1346t/26.Report.Sender.t                       (Wstat: 65280 Tests: 1 Failed: 1)
1347  Failed test:  1
1348  Non-zero exit status: 255
1349  Parse errors: No plan found in TAP output
1350Files=20, Tests=923, 15 wallclock secs ( 0.22 usr  0.03 sys +  4.49 cusr  0.67 csys =  5.41 CPU)
1351Result: FAIL
1352The command "HARNESS_IS_VERBOSE=1 prove -v -Ilib t" exited with 1.
````

**Comment by marcbradshaw** [comment]
Looks like the problem was the wrong config being loaded. Commit added to force the test config for that test.

I think the config system needs another look, seems weird being able to have different config in each object, and each one loading its own config from disk each time. But that's for another MR. Was there any reason behind the current system?

**Comment by msimerson** [comment]
> I think the config system needs another look, seems weird being able to have different config in each object, and each one loading its own config from disk each time. But that's for another MR. Was there any reason behind the current system?

One of the features is that if you want a config file to overload the default config file, you can see it being done in several of the test files:

- 03.Base.t:eval { $base->config('t/mail-dmarc.ini'); };
- 04.PurePerl.t:$dmarc->config('t/mail-dmarc.ini');
- 11.Report.Store.t:$dmarc->report->store->backend->config('t/mail-dmarc.ini');
- 17.Report.Aggregate.Schema.t:$store->config('t/mail-dmarc.ini');
- 22.Report.Send.SMTP.t:$smtp->config('t/mail-dmarc.ini');
- 25.Report.Receive.t:$recv->config('t/mail-dmarc.ini');

**Pull request merged**
